src/OpenTelemetry.Exporter.Geneva/Metrics/TlvMetricExporter.cs

```diff
@@ -407,9 +407,18 @@ private static void SerializeMonitoringAccount(string monitoringAccount, byte[]
 
 #if EXPOSE_EXPERIMENTAL_FEATURES
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    private static void SerializeExemplars(Exemplar[] exemplars, byte[] buffer, ref int bufferIndex)
+    private static void SerializeExemplars(ReadOnlyExemplarCollection exemplars, byte[] buffer, ref int bufferIndex)
     {
-        if (exemplars.Length > 0)
+        var exemplarsCount = 0;
+        foreach (ref readonly var exemplar in exemplars)
+        {
+            if (exemplar.Timestamp != default)
```

Comment:
* The old code would give you back Exemplars in the array which may not have been updated. That's why this `Timestamp != default` check is here. The SDK now won't ever give those back. So you can remove this `Timestamp` check and just increment the count.

* Below you have another `foreach (var exemplar in exemplars)` block (and `Timestamp` check). Make that second block also use `ref` (`foreach (ref readonly var exemplar in exemplars)`) and remove the `Timestamp` check.

* This code is written at the moment to loop once in order to get the count. Then it loops again to write out the data if there were exemplars detected. Could we fold that into one loop? Something like...

```csharp
var payloadTypeStartIndex = bufferIndex + 2;
var exemplarCount = 0;

foreach (ref readonly var exemplar in exemplars)
{
   if (exemplarCount++ == 0)
   {
      MetricSerializer.SerializeByte(buffer, ref bufferIndex, (byte)PayloadType.Exemplars);
      bufferIndex += 2;
      MetricSerializer.SerializeByte(buffer, ref bufferIndex, 0); // version <- is this version? seems like payload length :shrug:
   }

   SerializeSingleExmeplar(exemplar, buffer, ref bufferIndex);
}

if (exemplarCount > 0)
{
      var payloadTypeLength = (ushort)(bufferIndex - payloadTypeStartIndex);
      MetricSerializer.SerializeUInt16(buffer, ref payloadTypeStartIndex, payloadTypeLength);
}

Reply:
@codeblanch - updated. We do need to have the exemplarCount before writing the exemplars. [This](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/d9d591e73f32315eb008216f8b76ad1088ed7f6b/src/OpenTelemetry.Exporter.Geneva/Metrics/TlvMetricExporter.cs#L428) can take variable # of bytes I think.

Reply:
Yeah, we need to know the `exemplarCount` beforehand. We can try to backfill the count after serializing the individual exemplars but it's not as straightforward since `exemplarCount` is serialized using **varint** encoding.